### PR TITLE
💻 Rerouting app & adding events list page and events detail page

### LINKED ISSUE #36: Create events list and details page

### PR DESCRIPTION
**Pull request by lesliexin**
## Purpose
Closes #36. As a user, when I navigate to the Event lists page, I should be able to view a list of events and click on them to see their details. 

## Approach
_What approach are you choosing? Why?_
Pulling out shared hook as the search for both events and clubs uses the same logic.

## Screenshots
ezgif com-gif-maker (3)

## Learning
_Describe the research stage_
Learned about how to actually organize files w/ next.js and that the `pages` directory shouldn't house any component logic.

#### Blog Posts
- [How to Pull Request](https://github.com/flexyford/pull-request) Github Repo with Learning focused Pull Request Template.

### CODE REVIEW

**Comment by lesliexin** [comment]
Going to try and pull the search out into a shared hook

**Comment by jeffzh4ng** [comment]
Can you prepend the 💻 emoji to the PR title?

File: client/components/clubs/ClubList.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        <ClubCard
+          club={{
+            id: club.id,
+            name: club.name,
+            description: club.description,
+          }}
+        />
+      );
+    })
+  }
```
Comment (jeffzh4ng) [inline client/components/clubs/ClubList.tsx:27]: Do you think

```
export default function ClubList({clubListData}: ClubListProps) {
  const [filteredClubs, setSearchValue] = useSearch("", clubListData);

   return (
      <div>
        <h1>Club List</h1>
        <div>
          <input onChange={e => setSearchValue(e.target.value)}/>
        </div>
        { clubsList(filteredClubs) }
      </div>
    );
}

const clubsList = (filteredClubs: 🤷‍♂️) => {
  return filteredClubs.map((club: SearchItem | { item: SearchItem }) => {
    if ("item" in club) {
      club = club.item as SearchItem;
    }
    return (
      <ClubCard
        club={{
          id: club.id,
          name: club.name,
          description: club.description,
        }}
      />
    );
  })
```

File: client/components/events/EventList.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  return (
+    <div>
+      <h1>Event List</h1>
+      <div>
+        <input onChange={e => setSearchValue(e.target.value)}/>
+      </div>
+      { eventsList() }
+    </div>
```
Comment (jeffzh4ng) [inline client/components/events/EventList.tsx:36]: same suggestion as the one in  `<ClubsList>` (moving eventsList out)

File: client/components/hooks/UseSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      keys: ["name"],
+      ignoreLocation: true, // doesn't weight string location
+      threshold: 0.3 // 0=strict, 1=loose
+    })
+  }, [formattedSearchItems])
+
+  const filteredSearchItems: (Fuse.FuseResult<SearchItem> | SearchItem)[] = searchValue ? fuse.search(searchValue) : formattedSearchItems;
+  
+  return [filteredSearchItems, setSearchValue];
+}
```
Comment (jeffzh4ng) [inline client/components/hooks/UseSearch.tsx:37]: idk what's more beautiful to me this clean hook or your recent painting 😍

**Review by jeffzh4ng — CHANGES_REQUESTED** [review]

File: client/components/events/EventCard.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import { useRouter } from "next/router";
+import { Id, EventInfo } from "../../pages/_app";
+
+type Event = EventInfo & { id: Id };
+
+interface EventCardProps {
+  event: Event;
+}
+
+const ClubCard = ({event}: EventCardProps) => {
```
Comment (xujustinj) [inline client/components/events/EventCard.tsx:11]:
```suggestion
const EventCard = ({event}: EventCardProps) => {
```

File: client/components/events/EventCard.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+  return (
+    <div onClick={handleClick}>
+      <h1>{name}</h1>
+      <p>{description}</p>
+    </div>
+  );
+};
+
+export default ClubCard;
```
Comment (xujustinj) [inline client/components/events/EventCard.tsx:27]:
```suggestion
export default EventCard;
```

File: client/components/events/index.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export { default } from "./EventList";
```
Comment (xujustinj) [inline client/components/events/index.ts:1]: I wonder if it'd be better to make these exports explicit so we know exactly what we're exporting

File: client/components/hooks/UseSearch.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import Fuse from 'fuse.js';
+
+export type SearchItemData = {
+  name: string,
+  description: string
+}
+
+export type SearchItem = { id: number } & SearchItemData;
+
+type UseSearchReturn = [(Fuse.FuseResult<SearchItem> | SearchItem)[], Dispatch<SetStateAction<string>>];
```
Comment (xujustinj) [inline client/components/hooks/UseSearch.tsx:11]:
```suggestion
type SearchResult = (Fuse.FuseResult<SearchItem> | SearchItem)[];
type UseSearchReturn = [SearchResult, Dispatch<SetStateAction<string>>];
```

also is `(Fuse.FuseResult<SearchItem>[] | SearchItem[])` a more accurate type?

File: client/components/hooks/UseSearch.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+  const fuse = useMemo(() => {
+    return new Fuse(formattedSearchItems, {
+      keys: ["name"],
+      ignoreLocation: true, // doesn't weight string location
+      threshold: 0.3 // 0=strict, 1=loose
+    })
+  }, [formattedSearchItems])
+
+  const filteredSearchItems: (Fuse.FuseResult<SearchItem> | SearchItem)[] = searchValue ? fuse.search(searchValue) : formattedSearchItems;
```
Comment (xujustinj) [inline client/components/hooks/UseSearch.tsx:34]:
```suggestion
  const filteredSearchItems: SearchResult = searchValue ? fuse.search(searchValue) : formattedSearchItems;
```

**Review by xujustinj — CHANGES_REQUESTED** [review]
Is it intended that the homepage will be Events?

File: client/components/hooks/UseSearch.tsx (hunk 7 above)
Comment (lesliexin) [reply client/components/hooks/UseSearch.tsx:11]: [seems like the array brackets should be outside the union type ](https://www.codecademy.com/learn/learn-typescript/modules/learn-typescript-union-types/cheatsheet)

Comment (xujustinj) [reply client/components/hooks/UseSearch.tsx]: The distinction here is that the array is either 100% `SearchItem`s (i.e. `formattedSearchItems`) or 100% `Fuse.FuseResult<SearchItem>`s (i.e. `fuse.search(searchValue)`).

`(Fuse.FuseResult<SearchItem> | SearchItems)[]` is a slightly looser type that lets you have both in the same array

image

Comment (lesliexin) [reply client/components/hooks/UseSearch.tsx]: Ohhh I see, that makes sense :)

Comment (lesliexin) [reply client/components/hooks/UseSearch.tsx]: Getting error that the union type is not callable :( https://stackoverflow.com/questions/49365114/typescript-fails-type-resolution-on-union-type. Maybe the way we had it before is okay?

Comment (xujustinj) [reply client/components/hooks/UseSearch.tsx]: 😩 TypeScript

Yeah the way it was before is best then

**Review by xujustinj — APPROVED** [review]
🚀

**Review by jeffzh4ng — DISMISSED** [review]

**Review by jeffzh4ng — APPROVED** [review]

**Review by xujustinj — APPROVED** [review]

**Pull request merged**